### PR TITLE
FIX: Rules widget_map can change during iteration

### LINKED ISSUE #643: Rules widget_map changes during iteration
https://github.com/slaclab/pydm/blob/daf14f410452b8ef2cccf5ed62dd92bcc46283c8/pydm/widgets/rules.py#L211

Fix should be to wrap `self.widget_map` like so: `for widget_ref in list(self.widget_map):`

### PR DESCRIPTION
**Pull request by klauer**
Closes #643 - can happen when iterating over keys in `run`.

### CODE REVIEW

**Comment by hhslepicka** [comment]
@klauer I pushed some changes... do you mind taking a look at them?

File: pydm/widgets/rules.py [hunk 1]
```diff
@@ -292,7 +295,13 @@ def calculate_expression(self, widget_ref, rule):
         -------
         None
         """
-        rule['calculate'] = False
+        def safe_reset_calculate():
+            try:
+                self.widget_map[widget_ref][idx]['calculate'] = False
+            except (Exception, RuntimeError):
```
Comment (klauer) [inline pydm/widgets/rules.py:301]: `isinstance(RuntimeError(), Exception) = True` - unnecessary to check for both.

**Comment by klauer** [comment]
I think it's probably good - but to be honest without spending a considerable amount of time looking into the locking, I can't really say for certain.

One thing to keep in mind is that `.get()` and `.pop()` and `d[key] = value` are atomic (i.e., if used in isolation, there's no locking required because of the GIL), but I'm fairly certain things like `items()` are not (and locking around them as you're doing is necessary).

**Comment by hhslepicka** [comment]
I tested this with @tashasummers and it is working now without error messages.  As soon as tests are passing we are safe to merge.

**Pull request merged**